Week 06/Lecture 11/Assignment 01/employee/src/main/java/com/example/employee/service/impl/EmployeeServiceImpl.java

```diff
@@ -0,0 +1,177 @@
+package com.example.employee.service.impl;
+
+import com.example.employee.criteria.EmployeeSearchCriteria;
+import com.example.employee.dto.EmployeeDTO;
+import com.example.employee.dto.SalaryDTO;
+import com.example.employee.dto.TitleDTO;
+import com.example.employee.entity.*;
+import com.example.employee.repository.EmployeeRepository;
+import com.example.employee.repository.SalaryRepository;
+import com.example.employee.repository.TitleRepository;
+import com.example.employee.service.EmployeeService;
+import com.example.employee.specification.EmployeeSpecification;
+import jakarta.transaction.Transactional;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.springframework.stereotype.Service;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+@Service
+@Transactional
+public class EmployeeServiceImpl implements EmployeeService {
+
+    private final EmployeeRepository employeeRepository;
+
+    @Autowired
+    public EmployeeServiceImpl(EmployeeRepository employeeRepository) {
+        this.employeeRepository = employeeRepository;
+    }
+
+    /**
+     * Get all employees with pagination
+     * @param pageable pagination information
+     * @return page of employee DTO
+     */
+    @Override
+    @Transactional
+    public Page<EmployeeDTO> getAllEmployees(Pageable pageable) {
+        Page<Employee> employees = employeeRepository.findAll(pageable);
```

Comment:
Should notice when use findAll default not query

Reply:
Thank you for the feedback, I've added `@Query` and `fetch` to avoid the N+1 query issues, the following is one of my updated code:

```java
@Query("SELECT e FROM Employee e LEFT JOIN FETCH e.salaries LEFT JOIN FETCH e.titles LEFT JOIN FETCH e.deptEmps LEFT JOIN FETCH e.deptManagers")
    Page<Employee> findAll(Pageable pageable);
```